### DISCUSSION TITLE
etcdutl Snapshot Status: TotalKey Calculation Inquiry

### DISCUSSION THREAD

**Discussion by jxustc** [etcd-io/etcd#19012] [Q&A] [answered]
## Problem Description
When using the `etcdutl snapshot status` command, I've observed the following characteristics of TotalKey calculation:
- Current implementation counts keys across all buckets in the bbolt database
- Includes keys beyond visible etcd key-value data
- Key counts in other buckets may vary with different etcd versions

## Case
<img width="978" alt="image" src="https://github.com/user-attachments/assets/dd2093de-1357-476a-87b0-a3d5de2ed21d">

![image](https://github.com/user-attachments/assets/7878cd3a-28ec-4fbf-b586-033beb4d0d3a)

During  testing, I discovered that the TotalKey calculation includes keys from various internal buckets beyond the main key-value store:
### Identified Included Keys
- In `meta` bucket:
  - `term`
  - `scheduledCompactRev`
  - `confState`
  - ...
- In `auth` bucket:
  - `authRevision`
- Keys from other internal buckets

## Specific Questions
1. What is the precise definition of TotalKey?
2. Is the current calculation method appropriate?
3. Should the count be limited to etcd key-value entries?

## Potential Implementation Approach

```golang
func (s *v3Manager) Status(dbPath string) (ds Status, err error) {
	...
	if err = db.View(func(tx *bolt.Tx) error {
		...
		c := tx.Cursor()
		for next, _ := c.First(); next != nil; next, _ = c.Next() {
			b := tx.Bucket(next)
			if b == nil {
				return fmt.Errorf("nil bucket: %q", string(next))
			}
			_, err = h.Write(next)
			if err != nil {
				return fmt.Errorf("cannot hash bucket name: %q err: %w", string(next), err)
			}

			iskeyb := (bytes.Equal(next, schema.Key.Name()))
			if err = b.ForEach(func(k, v []byte) error {
				_, err = h.Write(k)
				if err != nil {
					return fmt.Errorf("cannot hash bucket key: %q err: %w", k, err)
				}
				_, err = h.Write(v)
				if err != nil {
					return fmt.Errorf("cannot hash bucket key: %q value: %q err: %w", k, v, err)
				}
				if iskeyb {
					var rev mvcc.Revision
					rev, err = bytesToRev(k)
					if err != nil {
						return fmt.Errorf("cannot parse revision key: %q err: %w", k, err)
					}
					ds.Revision = rev.Main

					var kv mvccpb.KeyValue
					err = kv.Unmarshal(v)
					if err != nil {
						return fmt.Errorf("cannot unmarshal value, key: %q value: %q err: %w", k, v, err)
					}
                                         // move to here 
                                         ds.TotalKey++
				}
                                 // remove 
				ds.TotalKey++
				return nil
			}); err != nil {
				return fmt.Errorf("error during bucket key iteration, name: %q err: %w", string(next), err)
			}
		}
		return nil
	}); err != nil {
		return ds, err
	}
...
}
```

**Besides this solution, could we possibly add a field to count the number of non-built-in key-value pairs in etcd?**

## Proposed Improvement

If appropriate, I am willing to submit a PR to implement this change.

**Comment by ahrtr** [comment]
Sorry for the late response, please see my answer in https://github.com/etcd-io/etcd/issues/19253#issuecomment-2607794507

**Comment by serathius** [comment, marked as answer]
With the issue being confirmed and discussed in https://github.com/etcd-io/etcd/issues/19253 can close this.
